backend/model_server/encoders.py

```diff
@@ -25,6 +35,100 @@
 _RERANK_MODELS: Optional[list["CrossEncoder"]] = None
 
 
+class EmbeddingProvider(Enum):
+    OPENAI = "openai"
+    COHERE = "cohere"
+    VOYAGE = "voyage"
+    VERTEX = "vertex"
+
+
+class CloudEmbedding:
+    def __init__(self, api_key: str, provider: str, model: str | None = None):
+        self.api_key = api_key
+        self.model = model
+        try:
+            self.provider = EmbeddingProvider(provider.lower())
+        except ValueError:
+            raise ValueError(f"Unsupported provider: {provider}")
+        self.client = self._initialize_client()
+
+    def _initialize_client(self) -> Any:
+        if self.provider == EmbeddingProvider.OPENAI:
+            return openai.OpenAI(api_key=self.api_key)
+        elif self.provider == EmbeddingProvider.COHERE:
+            return CohereClient(api_key=self.api_key)
+        elif self.provider == EmbeddingProvider.VOYAGE:
+            return voyageai.Client(api_key=self.api_key)
+        elif self.provider == EmbeddingProvider.VERTEX:
+            credentials = service_account.Credentials.from_service_account_info(
+                json.loads(self.api_key)
+            )
+            project_id = json.loads(self.api_key)["project_id"]
+            vertexai.init(project=project_id, credentials=credentials)
+            return TextEmbeddingModel.from_pretrained(
+                self.model or "text-embedding-004"
+            )
+
+        else:
+            raise ValueError(f"Unsupported provider: {self.provider}")
+
+    def encode(self, texts: list[str], model_name: str | None) -> list[list[float]]:
+        return [self.embed(text, model_name) for text in texts]
+
+    def embed(self, text: str, model: str | None = None) -> list[float]:
+        # logger.debug(f"Embedding text with provider: {self.provider}")
+        if self.provider == EmbeddingProvider.OPENAI:
+            return self._embed_openai(text, model or "text-embedding-ada-002")
+        elif self.provider == EmbeddingProvider.COHERE:
+            return self._embed_cohere(text, model)
+        elif self.provider == EmbeddingProvider.VOYAGE:
+            return self._embed_voyage(text, model)
+        elif self.provider == EmbeddingProvider.VERTEX:
+            return self._embed_vertex(text, model)
+        else:
+            raise ValueError(f"Unsupported provider: {self.provider}")
+
+    def _embed_openai(
+        self, text: str, model: str = "text-embedding-ada-002"
+    ) -> list[float]:
+        logger.debug(f"Using OpenAI embedding with model: {model}")
+        response = self.client.embeddings.create(input=text, model=model)
+        return response.data[0].embedding
+
+    def _embed_cohere(self, text: str, model: str | None) -> list[float]:
+        if model is None:
+            model = "embed-english-v3.0"
+
+        logger.debug(f"Using Cohere embedding with model: {model}")
+        response = self.client.embed(
+            texts=[text], model=model, input_type="search_document"
```

Comment:
A lot of these need to have the input type and prefix correctly, ask me if you have questions

Reply:
Correct approach would be to have the different models use their defaults, a prefix can be provided as an override but generally we would just use the flag for document vs query.